### ISSUE TITLE
[Ventura M2 Pro] Dim HDR when playing any HDR content (HDR10, 10+, Vision)

### ISSUE THREAD

**Issue by GameTL** [iina/iina#4214]
**System and IINA version:**
- macOS Ventura 13.2
- IINA 1.3.1 Build 133, release version

**Expected behavior:**
HDR plays at the right brightness when playing a video

**Actual behavior:**
the brightness of video playing is more like 1 bar of brightness when my display is at "full brightness"(not watchable), other elements and windows on the display is very bright just IINA widows that are dim.
The dimming is gone when HDR toggle is switched off. 
When the video is pause and the focus is on other app the video was back to a very bright image, might just be SDR at full brightness, not sure. This happens everytime 

***details***
I have tried the fixed that worked for M2 Air but I have the M2 Pro that was just released, so. I guess it is different. 
exporting the video to mp4 with Subler and play on Quicktime did trigger the HDR mode.

Thanks for this app, definitely better and more modern than VLC

**Comment by low-batt** [comment]
Just yesterday I was wondering what the behavior was on the new MacBook Pros with the M2 chip.

The "fix" you tried that worked on the MacBook Air M2 was it this one mentioned in issue #3906?:
> Go to IINA menubar -> Video -> Crop -> Custom... -> Press the Cancel button in the right bottom corner

Does this only occur in full screen mode, or is the same dimming experienced when IINA is playing a HDR video in a window?

Here is what I know about this issue…

First the situation with the MacBook Pros that have the M1 chip. With these Macs HDR was working under macOS Monterey. The problem appeared with macOS 13.0. The dimming only happens with HDR videos when playing in full screen mode. The dimming does not occur until the OSC disappears.

In general, when a software component is updated and a new problem appears in an application the fault lies with the component that was updated. However this is just a general rule. It could be that:
- There was an existing error in another component that changes in the updated component brought out
- There are changes in the updated component that are intentionally not backward compatible and require other components to be updated

Ventura updates [Metal](https://developer.apple.com/metal/) to version 3. So we know for certain there have been major changes related to graphics in this release of macOS.

This report about the M2 MacBook Pro prompted me to watch the WWDC22 video [Discover Metal 3](https://developer.apple.com/videos/play/wwdc2022/10066/). I did not hear anything that would explain the dimming, but graphics is outside of my area of expertise.

IINA is using [OpenGL](https://www.opengl.org/) which has been deprecated by Apple. So one worry is that Apple has broken OpenGL support. The [mpv](https://mpv.io/) project is working on a new generation renderer that will use Metal, but it is still being developed.

As the dimming on the M1 MacBook Pros does not occur when the OSC is displayed a workaround was added to IINA 1.3.1 that adds a tiny transparent view to the displayed content so there is always more than one view being displayed. I have confirmed this ughly workaround is still needed under macOS 13.2.

In issue #3906 there is a comment that indicates the M1 MacBook Air also has this same problem under Ventura.

On to the situation with the M2 MacBook Air. In issue #3906 you will see a good bit of confusion in the comments due to there being more than one problem causing HDR to look dim. Adding to the confusion is that the MacBook Air does not have a XDR display which makes HDR ([EDR](https://developer.apple.com/videos/play/wwdc2021/10161/)) support much more sensitive to the environment. Of particular note is that the problem on the M2 MacBook Air was seen under macOS Monterey and not just Ventura.

With this report of the problem occurring with the M2 MacBook Pro which has a XDR display we can eliminate the type of screen as a factor.

All of this suggests to me the root cause is in macOS and a fix from Apple is needed.

**Comment by SongTonyLi** [comment]
Thanks for mentioning the hacky solution mentioned in issue #3906. This solution works for my MacBook pro 2022 with m2 chip. HDR video is no longer dark anymore. However, I can literally use my eyes to tell that the video is not HDR anymore. Also, if I quit IINA, and restart it, then the same HDR video is completely dark again.

**Comment by low-batt** [comment]
Got me to run a few more tests today. This is on my MacBookPro18,2 with the M1 chip. I had forgotten one of the symptoms of this problem as I rarely encounter it. Playing an HDR video in full screen when I pause playback usually the screen may or may not dim. Rarely it starts flashing. The screen will brighten up momentarily displaying the paused frame with the proper luminance then rapidly back to dark.

I also confirmed the workaround of starting to crop the video and then canceling also prevents the dimming on the M1. Unlike the report above for the M2 the crop workaround does not suppress HDR on the M1. Disturbing this works as a workaround. You would think after canceling the operation IINA would function exactly as before you started the canceled crop operation. 

These tests were run with IINA 1.3.0 modified to use mpv 0.35.0.

**Comment by GameTL** [comment]
Right, sorry it took so long to reply back. the dimming issue on the XDR display on the M2 pro 16 inch is present whenever the video is pause or playing. No matter if the video playing is in full screen or windowed.

**Comment by mxsrm** [comment]
> Right, sorry it took so long to reply back. the dimming issue on the XDR display on the M2 pro 16 inch is present whenever the video is pause or playing. No matter if the video playing is in full screen or windowed.

I can confirm that behavour on a new M2 MBP 14" with M2 Pro processor.

**Comment by xqm32** [comment]
> > Right, sorry it took so long to reply back. the dimming issue on the XDR display on the M2 pro 16 inch is present whenever the video is pause or playing. No matter if the video playing is in full screen or windowed.
> 
> I can confirm that behavour on a new M2 MBP 14" with M2 Pro processor.

I have the same problem on the MacBook Air (M2 chip)

**Comment by alessandronoiato** [comment]
Same here, on a MacBook Pro M2 Max 2023

**Comment by low-batt** [comment]
I just confirmed the dimming in full screen mode when playing a HDR video that appeared for Macs with the M1 chip with the release of macOS Ventura is **not fixed** in the just released macOS 13.3. IINA 1.3.1 will not exhibit this problem because an awful hackaround was added in 1.3.1. That workaround is still working with macOS 13.3.

I do not have a Mac with the M2 chip, so someone with a Mac that has that chip will have to report on the situation with the M2 and macOS 13.3. As the M1 problem has not been fixed I'm expecting the M2 behavior has not changed.

**Comment by dpekez** [comment]
Just tested this on my M2 Pro MacBook Pro. macOS 13.3 and IINA 1.3.1 Build 133. The problem still is the same. Here are some situation where HDR playback is as intended though:

- first few seconds when starting the playback
- while resizing the player window
- first few seconds after enabling HDR

**Comment by low-batt** [comment]
Thanks for the report. Does the M2 workaround mentioned in issue #3906 still work?:
> Go to IINA menubar -> Video -> Crop -> Custom... -> Press the Cancel button in the right bottom corner

**Comment by dpekez** [comment]
Ah, thanks for pointing that out. Yes, HDR works that way!

**Comment by low-batt** [comment]
It is so strange that HDR starts working after that sequence. This is one of the reasons we suspect the problem is in the macOS graphics system.

**Comment by ibLeDy** [comment]
I just discovered that if I open a video in VLC and leave it paused, then the HDR will work correctly in IINA 🤔 

- IINA `1.3.1`
- VLC `3.0.18`
- MacOS Ventura `13.4`
- M2 Pro MacBook Pro 14"

Also, just to add to the other comments, HDR works fine for a split second every time I move the cursor

**Comment by svobs** [comment]
@ibLeDy it sounds like this issue may be fixed in the latest [nightly build](https://nightly.iina.io/) - can you try that?

**Comment by ibLeDy** [comment]
@svobs Just tried it, HDR is working fine for me!

**Comment by svobs** [comment]
Great! The core team is supposedly close to packaging a new official release, and hopefully they will finish it up soon. There have been several critical bugs fixed since the last release.

**Comment by low-batt** [comment]
Work on the next release is at the point where upgraded dependencies are being built and tested. So we are getting close to a release. Progress is being tracked in the pinned issue.
